### PR TITLE
Passwordbox Reveal Button

### PR DESCRIPTION
**Pull request by faxe1008**
This PR adds a button to the PasswordBox which allows for revealing of the typed password.
The button is only shown when the property show_reveal_button is true. It's default is false to not unintentionally alter existing GUIs.

Unrevealed state:
image

Revealed state:
image

I added this because of some password related stuff I am working on and it bugged me during developing that the password can not be shown :^). The eye icon is not the prettiest but I could not come with something better.

### CODE REVIEW

File: Userland/Libraries/LibGUI/TextEditor.cpp [hunk 1]
```diff
@@ -2066,9 +2066,10 @@ void TextEditor::set_should_autocomplete_automatically(bool value)
...
 }
 
-void TextEditor::set_substitution_code_point(u32 code_point)
+void TextEditor::set_substitution_code_point(Optional<u32> code_point)
 {
-    VERIFY(is_unicode(code_point));
+    if (code_point.has_value())
+        VERIFY(is_unicode(code_point.value()));
     m_substitution_string_data.clear();
     m_substitution_code_point = code_point;
```
Comment (linusg) [inline Userland/Libraries/LibGUI/TextEditor.cpp:2074]:
```suggestion
    m_substitution_code_point = move(code_point);
```

File: Userland/Libraries/LibGUI/TextBox.cpp [hunk 2]
```diff
@@ -5,8 +5,9 @@
  * SPDX-License-Identifier: BSD-2-Clause
  */
 
+#include <LibGUI/Painter.h>
 #include <LibGUI/TextBox.h>
-
+#include <LibGfx/Palette.h>
```
Comment (linusg) [inline Userland/Libraries/LibGUI/TextBox.cpp:10]: Please keep the empty line after the includes

File: Userland/Libraries/LibGUI/TextBox.cpp [hunk 3]
```diff
@@ -71,11 +72,59 @@ void TextBox::add_input_to_history(String input)
...
 {
-    set_substitution_code_point('*');
+    set_substitution_code_point(password_box_substitution_code_point);
     set_text_is_secret(true);
+    REGISTER_BOOL_PROPERTY("show_reveal_button", is_showing_reveal_button, set_show_reveal_button);
+}
+
+Gfx::IntRect PasswordBox::reveal_password_button_rect() const
+{
+    constexpr i32 button_box_padding = 3;
```
Comment (linusg) [inline Userland/Libraries/LibGUI/TextBox.cpp:87]: Seems like this is actually a margin?

File: Userland/Libraries/LibGUI/TextBox.cpp [hunk 4]
```diff
@@ -71,11 +72,59 @@ void TextBox::add_input_to_history(String input)
...
+    if (is_showing_reveal_button()) {
+        auto button_rect = reveal_password_button_rect();
+
+        Painter painter(*this);
+        painter.add_clip_rect(event.rect());
+
+        auto icon_color = palette().button_text();
+        if (substitution_code_point().has_value()) {
+            icon_color = palette().disabled_text_front();
+        }
```
Comment (linusg) [inline Userland/Libraries/LibGUI/TextBox.cpp:105]: No curly braces needed here

**Review by linusg — COMMENTED** [review]
Would be nicer if it was an actual button with focus behavior IMO, this one can't be keyboard activated

File: Userland/Libraries/LibGUI/TextBox.cpp (hunk 3 above)
Comment (faxe1008) [reply Userland/Libraries/LibGUI/TextBox.cpp]: True, always get those mixed up :^)

**Comment by faxe1008** [comment]
> Would be nicer if it was an actual button with focus behavior IMO, this one can't be keyboard activated

I first wanted to make it a button, but I looked around in LibGUI and saw that ColorInput also has no dedicated button. 
Making it "tabable" would is nice but I wonder if that might lead to unintended password reveals when people are not expecting it to be a tabable thing. 

Sometime tomorrow I will try a version with a button. Do you suggest the button to sit flush with the border of the textbox?

**Review by linusg — APPROVED** [review]
Might be fine line this actually - I checked the UX on my machine (GNOME), and it's not focusable either. Perhaps there's a magic key to trigger it, but let's do this as a first iteration :+1:

**Pull request merged**